### PR TITLE
Add API endpoint to list mapped task instances

### PR DESCRIPTION
**Pull request by norm**
As part of AIP-42, a new API endpoint is needed that lists mapped task instances.

Screenshot of built docs — it cuts off, but the parameters should be identical to https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/get_task_instances

Screenshot 2022-03-22 at 15 30 47

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/api_connexion/openapi/v1.yaml [hunk 1]
```diff
@@ -1109,6 +1109,39 @@ paths:
         '404':
           $ref: '#/components/responses/NotFound'
 
+  /dags/{dag_id}/dagRuns/{dag_run_id}/taskInstances/{task_id}/list:
```
Comment (ashb) [inline airflow/api_connexion/openapi/v1.yaml]:
```suggestion
  /dags/{dag_id}/dagRuns/{dag_run_id}/taskInstances/{task_id}/listMapped:
```

**Comment by norm** [comment]
Needs to accept order and search params.

**Comment by norm** [comment]
> Could you add a screenshot of the OpenAPI rendered yaml file to the PR description? It would help to quickly see if the API has everything needed for people to use it.

Added, and I'll hopefully remember to do that in future without prompting :)

**Comment by uranusjr** [comment]
A task never mapped (either due to it not being a mapped task, or has not been run yet) would have one single task instance of `map_index=-1`. If the task is mapped, there will never be an instance with the -1 index. If the mapped length is zero, there will only be one task instance with index 0 and marked with state `SKIPPED`.

**Comment by bbovenzi** [comment]
I've been testing locally. Pagination and ordering seems to be working well!

**Comment by norm** [comment]
@uranusjr From my reading of the code, when the mapped length is 0, the state is set to skipped but the map_index is not altered to 0 before the return (also what my tests show when I debug them). Is this a bug in `mappedoperator.py` I should fix?

**Comment by uranusjr** [comment]
Hmm, I’m not sure. I remember we discussed marking the task instance as skipped, but not sure if we mentioned if that skipped task instance needs to be set to index 0. And this does not really matter anywhere anyway, so I think we can choose whatever makes things easier here.

**Comment by ashb** [comment]
I think that if we find no mapped TIs we should load the dag and check, because there it might be worth being able to differentiate between:

- mapped item was zero length, whole thing was skipped
- mapped item returned one item, and that one task execution chose to skip.

We could either load the DAG and check if `task.is_mapped`? Or just load all TIs (no map_index SQL filtering) and check things further only if we find an index of -1 (so that the "default" case of called on an expanded mapped index can run without extra queries, as loading the DAG is not free.)

**Comment by norm** [comment]
Since there would still be ambiguity in trying to figure it out from just the TI when both leaving it -1 and changing it to 0, checking `is_mapped` in the DAG seems the best way of resolving things. I'll try that.

**Review by uranusjr — APPROVED** [review]

File: tests/api_connexion/endpoints/test_task_instance_endpoint.py [hunk 2]
```diff
@@ -376,6 +376,16 @@ def test_should_return_404_for_mapped_endpoint(self, session):
...
 
+    def test_should_return_404_for_list_mapped_endpoint(self, session):
+        self.create_task_instances(session)
+        for index in ['0', '1', '2']:
+            response = self.client.get(
+                "/api/v1/dags/example_python_operator/dagRuns/TEST_DAG_RUN_ID/"
+                "taskInstances/print_the_context/listMapped",
+                environ_overrides={"REMOTE_USER": "test"},
+            )
+            assert response.status_code == 404
```
Comment (ashb) [inline tests/api_connexion/endpoints/test_task_instance_endpoint.py]: I don't think this for loop is needed/does anything now
```suggestion
        response = self.client.get(
            "/api/v1/dags/example_python_operator/dagRuns/TEST_DAG_RUN_ID/"
            "taskInstances/print_the_context/listMapped",
            environ_overrides={"REMOTE_USER": "test"},
        )
        assert response.status_code == 404
```

Comment (norm) [reply tests/api_connexion/endpoints/test_task_instance_endpoint.py]: Yep, that was thoughtless copy+paste on my part.

**Review by ashb — APPROVED** [review]

**Pull request merged**